### DISCUSSION TITLE
Base type for repo init and method return

### DISCUSSION THREAD

**Discussion by IlyaSemenov** [romeerez/orchid-orm#106] [Q&A]
Disclaimer: I know this is a weird question and there could be no good answer.

Anyhow, is there some reasonable base type that I can use for `Q` in the code below?

```ts
import { createRepo, DbTable } from 'orchid-orm'
import { Query } from 'pqb'

import { db } from '~/db'

const repo = createRepo(db.user, {
  queryMethods: {
    test: q => q.where({ id: 1 }),
  },
})

// What can I use for Q?
type Q = Query
type T = (query: Q) => Q

// Whatever I use for Q, this gives type error:
const modify: T = query => repo(query).test()

// However, this works if Q is Query or DbTable<any>:
const modify: T = query => query.where({ id: 1 })
```

- With `Q` set to `Query` there is a TS error when passing `query` to `repo` (it works in runtime, though)
- With `Q` set to `DbTable<any>` there is a TS error when returning the result of `test()` as it can't be cast to `DbTable<any>`.

Basically, my question is why `Query` can't be passed to `repo`?

Context: I reuse the type of `modify` between Orchid and Objection.js ([link](https://github.com/IlyaSemenov/graphql-orm/blob/fd33118555139a658d299140184025be004b8099/packages/graphql-orm/src/resolvers/table.ts#L24-L27)) and I'd like the input and output to be simply some kind of 'any query' for the time being.

**Comment by romeerez** [comment]
> why Query can't be passed to repo?

Because only `db.user` has the id of type number, so the `q.where({ id: 1 })` is valid for the user table, but may be invalid for other tables. Actually, this is not the real reason.

The real reason is that to make it work, the type of argument of `repo` is extended from the type of argument of `createRepo`:

```ts
// argument here is typeof db.user
const repo = createRepo(db.user)

// argument here is T extends typeof db.user
// apparently, Query doesn't extend it
// but DbTable<any> extends DbTable<user specifics>
repo(query)
```


The following seems to have no TS issues on my side:

```ts
const repo = createRepo(db.some, {
  queryMethods: {
    test: (q) => q.where({ id: 1 }),
  },
});

const modify = (query: DbTable<any>): Query => repo(query).test();

const one = await modify(db.some).order('id');
const two = await modify(db.other).order('id');
```

**Reply by IlyaSemenov** [reply]
> Because only db.user has the id of type number,

Sure thing, but `DbTable<any>` doesn't have `id` either, yet it works.

Similarly, `Query` doesn't have `id`, yet `query.where({ id: 1 })` works. So `Query` itself is generic enough to accept any where conditions, but somehow is not generic enough to be a repo argument?

In my example above, if I set `type Q = any`, it will work. My question was if there a **common base** between repo argument type and repo function return type other than `any`? Because from layman's point of view, both are 'queries' (in the sense that both have `.select`, `.where`, etc.), but somehow they don't seem to share any common base?

> The following seems to have no TS issues on my side:

Yes, it does, but that's not what I need. I have that function (`(query: Q) => Q`) as a generic somewhere else, so I'm thinking what type can I pass to it (and why `Query` is not enough).

**Comment by romeerez** [comment]
The argument of repo accepts `T extends typeof db.yourTable`. So if you created a repo for a user, `createRepo(db.user, ...)`, it won't allow you to use this repo for `db.product`, even though `db.user` and `db.product` both extend the `Query` interface.

> I have that function ((query: Q) => Q) as a generic somewhere else, so I'm thinking about what type can I pass to it (and why Query is not enough).

`createRepo` is expecting a concrete table, and you want to provide a generic type. Maybe this could work?

```ts
const genericRepo = createRepo(db.$queryBuilder, {
  // ...methods
})
```

And use the type `Query` for the function. `$queryBuilder` is a generic query object, it doesn't have a table name or a column. And it won't allow using `id` because there are no columns. To make any columns available, try `db.$queryBuilder as DbTable<any>`, to make only specific columns available, try `db.$queryBuilder as unknown as Query & { shape: { id: IntegerColumn } }`.
